components/trace/appmap/classmap/index.mjs

```diff
@@ -1,313 +1,105 @@
-              static: entity.static,
-              source: inline ? cutContent(entity.range) : null,
-              comment: printCommentArray(entity.comments),
-              labels: entity.labels,
-            },
-            ...entity.children.map(compileEntity),
-          ],
-        }
-      : {
-          ...entity,
-          children: entity.children.map(compileEntity),
-        };
-  return entities.map(compileEntity);
-};
-
 export const createClassmap = (configuration) => ({
-  closures: new Map(),
-  sources: [],
-  urls: new Set(),
-  naming: {
-    counter: createCounter(0),
-    separator: "-",
-  },
+  sources: new Map(),
   configuration,
 });
 
 export const addClassmapSource = (
   {
-    closures,
-    urls,
-    naming,
     configuration: {
       pruning,
       "function-name-placeholder": placeholder,
       repository: { directory },
     },
     sources,
   },
-  { url, content, inline, exclude, shallow },
+  { url, content, inline, exclude: exclusions, shallow },
 ) => {
-  assert(!urls.has(url), "duplicate source url", InternalAppmapError);
-  urls.add(url);
-  const relative = toRelativeUrl(url, directory);
-  assert(
-    relative !== null,
-    "could not extract relative url",
-    InternalAppmapError,
-  );
-  const context = { url, relative, shallow, inline, content, placeholder };
-  const exclusions = compileExclusionList(exclude);
-  const excluded_entities = [];
-  let entities = extractEstreeEntityArray(relative, content, naming).flatMap(
-    (entity) => excludeEntity(entity, null, exclusions, excluded_entities),
+  assert(!sources.has(url), "duplicate source url", InternalAppmapError);
+  sources.set(
+    url,
+    createSource(content, {
+      url,
+      directory,
+      placeholder,
+      pruning,
+      inline,
+      exclusions,
+      shallow,
+    }),
   );
-  for (const entity of excluded_entities) {
-    const { type } = entity;
-    if (type === "function") {
-      const { line, column } = entity;
-      closures.set(stringifyLocation(makeLocation(url, line, column)), null);
-    }
-  }
-  registerEntityArray(entities, context, closures);
-  if (pruning) {
-    entities = entities.flatMap(cleanupEntity);
-  }
-  sources.push({ context, entities });
 };
 
-export const getClassmapClosure = ({ closures }, location) => {
-  if (closures.has(location)) {
-    return closures.get(location);
-  }
-  const next_location = stringifyLocation(
-    incrementLocationColumn(parseLocation(location)),
-  );
-  if (closures.has(next_location)) {
-    logDebug(
-      "Had to increase column by one to fetch closure information at %j",
+export const lookupClassmapClosure = ({ sources }, location) => {
+  const base = getLocationBase(location);
+  if (sources.has(base)) {
+    return lookupSourceClosure(
+      sources.get(base),
+      getLocationPosition(location),
+    );
+  } else {
+    logWarning(
+      "Missing source file for closure at %s, threating it as excluded.",
```

Comment:
"treating" (or "threatening", but I doubt that's what you meant ;) )

Reply:
This closure is completely out of line. I'm *threatening* of exclusion! Rebased in 3c65c2448320c276ef2d297b3d39461c5848040c